### PR TITLE
Gracefully handle unknown message id's in handlePubrel

### PR DESCRIPTION
**Pull request by shylesh107**
This should handle PUBREL messages without throwing an exception for unknown message id's. 
See #415 for more details.

### CODE REVIEW

**Comment by mcollina** [comment]
Can you please add a unit test for this?

I think it's ok, but can anybody else have a look as well?

File: lib/client.js [hunk 1]
```diff
@@ -866,15 +866,10 @@ MqttClient.prototype._handlePubrel = function (packet, callback) {
...
-      return that.emit('error', err);
-    }
-
-    if ('pubrel' !== pub.cmd) {
+    if (!err && 'pubrel' !== pub.cmd) {
       that.emit('message', pub.topic, pub.payload, pub);
-      that.incomingStore.put(packet);
     }
-
+    that.incomingStore.put(packet);
```
Comment (mcollina) [inline lib/client.js]: I'm not really convinced by this, I think this should be a `incomingStore.del()` in the first place.

Comment (shylesh107) [inline lib/client.js]: Yea, I had that doubt as well.

**Comment by mcollina** [comment]
Any update on this?

**Comment by vingorius** [comment]
Tks @shylesh107 , Your code fix my problem.

**Comment by mcollina** [comment]
@vingorius would you like to send a PR with the change and a unit test?

**Comment by vingorius** [comment]
Ok @mcollina , i zipped two files.

first run `node mqtt_client_pr.js`, it will be ready to subscribe messages.
then run `bash ./pub_pr.sh` on command line. 
it will repeat sending message 100 times.

when at any time, stop receiving with `Ctr-C` on `node mqtt_client_pr.js` terminal window.

let  `bash ./pub_pr.sh`  send messages.

finally rerun  `node mqtt_client_pr.js` .
it throw error

```
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: missing packet
    at Store.get (/home/vingorius/projects/mqtttest/node_modules/mqtt/lib/store.js:93:8)
    at MqttClient._handlePubrel (/home/vingorius/projects/mqtttest/node_modules/mqtt/lib/client.js:905:22)
    at MqttClient._handlePacket (/home/vingorius/projects/mqtttest/node_modules/mqtt/lib/client.js:297:12)
    at process (/home/vingorius/projects/mqtttest/node_modules/mqtt/lib/client.js:242:12)
    at Writable.writable._write (/home/vingorius/projects/mqtttest/node_modules/mqtt/lib/client.js:252:5)
    at doWrite (/home/vingorius/projects/mqtttest/node_modules/mqtt/node_modules/readable-stream/lib/_stream_writable.js:347:64)
    at writeOrBuffer (/home/vingorius/projects/mqtttest/node_modules/mqtt/node_modules/readable-stream/lib/_stream_writable.js:336:5)
    at Writable.write (/home/vingorius/projects/mqtttest/node_modules/mqtt/node_modules/readable-stream/lib/_stream_writable.js:274:11)
    at Socket.ondata (_stream_readable.js:555:20)
    at emitOne (events.js:96:13)

FYI : mqtt@2.3.1
```[pr.zip](https://github.com/mqttjs/MQTT.js/files/741353/pr.zip)

**Comment by mcollina** [comment]
@vingorius can you please submit a pull request to this repository with this change and a unit test? Thanks.

**Comment by vingorius** [comment]
PR means a pull request. 
@mcollina , i wish i could, but code was @shylesh107 , he(she) might do that. 
I just fixed my problem using @shylesh107 amendment, i even did not understand what @shylesh107 code mean. sorry @mcollina .

**Comment by shylesh107** [comment]
@vingorius You are welcome and thanks for taking the time to add a unit test.

@mcollina I got a little bit of time today to look in to adding a test case for this. But the current test suite may not work for this case. 

## Test case
- New client connection to broker with opts  `{clientId:"testClient", clean:false}`
- `testClient` subscribes to topic "test" (QoS 2)
- Another `senderClient` starts publishing messages to topic "test" with QoS 2
- `testClient` starts receiving messages, after a while close the connection to the broker
- `senderClient` continues sending messages to broker
- connect the `testClient` again with same opts
- `testClient` should be able to receive all messages from broker without error (https://github.com/mqttjs/MQTT.js/issues/415)

@mcollina I am not able to create multiple clients in the test file. It's probable that I am doing something stupid here. Please let me know if the above mentioned test case is valid and doable using the current test suite. Appreciate any pointers from anyone.

Thank you.

**Comment by mcollina** [comment]
The best way to test this is to spin up a new net server, and use mqtt-connection directly. Reply to CONNECT with CONNACK, and send a PUBREL with whatever id.

**Comment by peppesilletti** [comment]
Hi @mcollina, I am developing a library that relies on the mqtt package and I am facing this same problem. I really need to solve it, because it's critical for the functioning of our application and I would like to know if I can help in any way?

**Comment by mcollina** [comment]
@peppesilletti the best way would be to start working on a unit test that triggers the problem (and fail), and then work on a fix from there. You can open the PR when you have the unit test ready, or when you have the bug fixed. I have no time to work on this directly atm.

**Comment by peppesilletti** [comment]
Ok @mcollina I will have a look, thank you!

**Pull request closed without merging**